go-controller/pkg/clustermanager/userdefinednetwork/template/net-attach-def-template.go

```diff
@@ -1,15 +1,175 @@
+			CNIVersion: cniVersion,
+			Type:       ovnK8sCNIOverlay,
+			Name:       networkName,
+		},
+		NADName:  nadName,
+		Topology: strings.ToLower(string(udn.Spec.Topology)),
+	}
+
+	switch udn.Spec.Topology {
+	case userdefinednetworkv1.NetworkTopologyLayer3:
+		cfg := udn.Spec.Layer3
+		netConfSpec.Role = strings.ToLower(string(cfg.Role))
+		netConfSpec.MTU = int(cfg.MTU)
+		netConfSpec.Subnets = layer3SubnetsString(cfg.Subnets)
+		netConfSpec.JoinSubnet = cidrString(renderJoinSubnets(cfg.Role, cfg.JoinSubnets))
+	case userdefinednetworkv1.NetworkTopologyLayer2:
+		cfg := udn.Spec.Layer2
+		netConfSpec.Role = strings.ToLower(string(cfg.Role))
+		netConfSpec.MTU = int(cfg.MTU)
+		netConfSpec.AllowPersistentIPs = cfg.IPAMLifecycle == userdefinednetworkv1.IPAMLifecyclePersistent
+		netConfSpec.Subnets = cidrString(cfg.Subnets)
+		netConfSpec.JoinSubnet = cidrString(renderJoinSubnets(cfg.Role, cfg.JoinSubnets))
+	case userdefinednetworkv1.NetworkTopologyLocalNet:
+		cfg := udn.Spec.LocalNet
+		netConfSpec.Role = strings.ToLower(string(cfg.Role))
+		netConfSpec.MTU = int(cfg.MTU)
+		netConfSpec.AllowPersistentIPs = cfg.IPAMLifecycle == userdefinednetworkv1.IPAMLifecyclePersistent
+		netConfSpec.Subnets = cidrString(cfg.Subnets)
+		netConfSpec.ExcludeSubnets = cidrString(cfg.ExcludeSubnets)
+	}
+
+	if err := util.ValidateNetConf(nadName, netConfSpec); err != nil {
+		return nil, err
+	}
+	if _, err := util.NewNetInfo(netConfSpec); err != nil {
+		return nil, err
+	}
+
+	cniNetConf := map[string]interface{}{
+		"cniVersion":       cniVersion,
+		"type":             ovnK8sCNIOverlay,
+		"name":             networkName,
+		"netAttachDefName": nadName,
+		"topology":         netConfSpec.Topology,
+		"role":             netConfSpec.Role,
+	}
+	if mtu := netConfSpec.MTU; mtu > 0 {
+		cniNetConf["mtu"] = mtu
+	}
+	if len(netConfSpec.JoinSubnet) > 0 {
+		cniNetConf["joinSubnets"] = netConfSpec.JoinSubnet
+	}
+	if len(netConfSpec.Subnets) > 0 {
+		cniNetConf["subnets"] = netConfSpec.Subnets
+	}
+	if len(netConfSpec.ExcludeSubnets) > 0 {
+		cniNetConf["excludeSubnets"] = netConfSpec.ExcludeSubnets
+	}
+	if netConfSpec.AllowPersistentIPs {
+		cniNetConf["allowPersistentIPs"] = netConfSpec.AllowPersistentIPs
+	}
+
+	return cniNetConf, nil
+}
+
+func renderJoinSubnets(role userdefinednetworkv1.NetworkRole, joinSubnetes []userdefinednetworkv1.CIDR) []userdefinednetworkv1.CIDR {
+	if role != userdefinednetworkv1.NetworkRolePrimary {
+		return nil
+	}
+
+	if len(joinSubnetes) == 0 {
+		return []userdefinednetworkv1.CIDR{types.UserDefinedPrimaryNetworkJoinSubnetV4, types.UserDefinedPrimaryNetworkJoinSubnetV6}
+	}
+
+	return joinSubnetes
+}
+
+func layer3SubnetsString(subnets []userdefinednetworkv1.Layer3Subnet) string {
+	var cidrs []string
+	for _, subnet := range subnets {
+		if subnet.HostSubnet > 0 {
+			cidrs = append(cidrs, fmt.Sprintf("%s/%d", subnet.CIDR, subnet.HostSubnet))
+		} else {
+			cidrs = append(cidrs, string(subnet.CIDR))
```

Comment:
I added this check to avoid the scenario when no `HostSubnet` is specified the CIDR string will have `/0` for example: `10.100.0.0/16/0`.

The prasing NAD code [1], make use of `NewNetInfo` [2], does what you say, it set default `HostSubnet`which seem reasonable when parsing an actual state.

For UDN NAD case, where we create them based on the user desired spec, the template code avoid overriding what the given spec, it use `NewNetInfo` just for validation. 
In case bad `SubnetPrefix` is spesifed, it will raise an error in the CR status.

WDYT?

[1] https://github.com/ovn-org/ovn-kubernetes/blob/e02ff8cbedfb6d7bf490d23d1b38fe6c5e62c565/go-controller/pkg/util/multi_network.go#L681
[2] https://github.com/ovn-org/ovn-kubernetes/blob/e02ff8cbedfb6d7bf490d23d1b38fe6c5e62c565/go-controller/pkg/util/multi_network.go#L663

Reply:
OK it might be helpful to add some comments on this function saying something like:
```
layer3SubnetsString converts the subnet object into the proper CIDR/hostSubnet format. If the provided hostSubnet is 0 or not specified, then it is not appended to the CIDR. Post-validation of the hostSubnet being > the provided CIDR prefix, as well as defaulting the hostSubnet will happen during NAD processing and not in this CRD validation.
```

Reply:
Or maybe my comment is not accurate because it looks like in renderCNINetworkConfig you do 	if _, err := util.NewNetInfo(netConfSpec); err != nil {
		return nil, err
	}

which will parse the subnets and make sure they are valid right?

Reply:
Right, it parse the NAD validate the subnets only, no defaults are set whatsoever.
I think we can drop last part, maybe somthing along these lines?
```
// layer3SubnetsString converts the subnet object into the proper CIDR/hostSubnet format.
// If the provided hostSubnet is 0 or not specified, then it is not appended to the CIDR to prevent from parsing invalid CIDR/HostSubnet string. 
```

Reply:
Done